bytecomp/translcore.ml

```diff
@@ -413,6 +416,16 @@ let specialize_primitive p env ty ~has_constant_constructor =
     | (Parraysetu Pgenarray, p1 :: _) -> Parraysetu(array_type_kind env p1)
     | (Parrayrefs Pgenarray, p1 :: _) -> Parrayrefs(array_type_kind env p1)
     | (Parraysets Pgenarray, p1 :: _) -> Parraysets(array_type_kind env p1)
+    | (Parraylength Paddrarray, [p])
```

Comment:
Why is this chunk needed?

Reply:
As evidenced by `testsuite/tests/translprim/array_spec.ml`, without this code we lose some type specialization: without -no-flat-float-array, the compiler will specialize some accesses from `gen` to `int`, while with the option they are not `gen` but `addr`, and the compiler doesn't specialize them. This block of code recovers this optimization in the -no-flat-float-array case.

Reply:
Where does this specialization from gen to int happen?  Woudl'nt this be the place to change to make it happen from addr to int as well?  (Sorry if I miss something obvious.)

Reply:
Well... it happens right here in this pattern matching.

Reply:
Ok, yes, sorry.  I would find it easier to follow if the cases were merged:

````ocaml
   | (Parraylength old_kind, [p])   -> Parraylength(cap_array_kind old_kind (array_type_kind env p))
````